Extended TensorMetadata with device property

bohnstingl opened the pull request:
This PR intends to extend the TensorMetadata with the device property

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @ydwu4

bohnstingl committed:
Extended TensorMetadata with device property

ezyang reviewed (changes requested):
The location of the new field is BC breaking because this is a NamedTuple and thus subsequent fields were reordered

bohnstingl committed:
Moved new property to the end of the positional arguments and introduced default None argument in order not to break BC

bohnstingl commented:
Thank you @ezyang for looking at this. I moved the new property to the end and also introduced `None` as a default argument to ensure BC. Would this be better?

ezyang reviewed (approved):
suuuuure why not

ezyang commented:
@pytorchbot merge

ezyang commented:
Hmm, I guess you need to shut up some tests too

bohnstingl committed:
Accepted BC change to TensorMetadata

bohnstingl commented:
Yes, I monitored these tests. However, they were failing with 
```
****** ERROR ******
An FX class that has been marked as backwards-compatible has experienced change in its public members. See the above exception context for more information. If this change was unintended, please revert it. If it was intended, check with the FX team to ensure that the proper deprecation protocols have been followed and subsequently --accept the change.
```
and I wasn't sure whether I am allowed to run with `--accept`. I did now and the CI tests pass

bohnstingl commented:
No, I don't. I just thought that adding this property could be helpful as in some HOPs we wanted to check whether the output of the operations is still on the same device, that's why we added it.
In the case that you mentioned, when creating the TensorMetadata by hand, I have no preference. One way would be to enhance the TensorMetadata with a dedicated __eq__ function that ignores `None` values?

ezyang commented:
So, another thing I could suggest to you, is instead of using TensorMetadata, use the fake tensor in meta["val"] instead, which has the device. Does that work here?
